### ISSUE TITLE
Doctrine validation with DoctrineOrmPropertyMetadataFactory

### ISSUE THREAD

**Issue by daviddlv** [api-platform/api-platform#375]
Hi,

I've got an entity (api resource) as :
```PHP
class User implements UserInterface
{
    /**
     * @var string
     * @ORM\Column(length=200)
     * @Groups({"user_output", "user_input"})
     * @Assert\NotBlank
     * @Assert\Length(min="3", max="20")
     * @Assert\Regex("/^[-_\.a-z0-9]+$/i")
     */
    private $username;

    /**
     * @return string|null
     */
    public function getUsername(): ?string
    {
        return $this->username;
    }

    /**
     * @param string|null $username
     * @return User
     */
    public function setUsername(?string $username): User
    {
        $this->username = $username;

        return $this;
    }
}
```

I've got a Frontend on Angular and when i make a request with form data : {username: null}
I've got the error : The type of the "username" attribute must be "string", "NULL" given.

This error is due to the "decorated" property (DoctrineOrmPropertyMetadataFactory) in the "SerializerPropertyMetadataFactory"

I don't want to put `@ORM\Column(length=200, nullable=true)`

How to disable this validation ?

**Comment by soyuka** [comment]
If it's not nullable how is it that you accept requests with null values? IMO it's working fine.

**Comment by daviddlv** [comment]
By default with Angular, if i don't fill the field, it's null.
I would have preferred to get validation error messages `@Assert\NotBlank`

**Comment by vincentchalamon** [comment]
@soyuka The front shouldn't know what fields are allowed to be nullable or not in the API, it just wants to send data to the API and get a response.

Here is a use case: I have an entity with 2 fields (`foo` & `bar`). `foo` cannot be null, but `bar` can. What can I send to the API to empty `bar` value and set it to null if the front cannot send `null` values to the API?

**Comment by soyuka** [comment]
In your case:

```
{ bar: null }
```

> Here is a use case: I have an entity with 2 fields (foo & bar). foo cannot be null, but bar can. What can I send to the API to empty bar value and set it to null if the front cannot send null values to the API?

If `username` can not be nullable I don't get what response you expect when you send `{username: null}`, it's invalid. I'd expect `400` status code.

The `DoctrineOrmPropertyMetadataFactory` is only used to identify identifiers if I'm not mistaken. Could we have a full stack trace?

**Comment by soyuka** [comment]
Angular version? 
<details>
<summary>wrote this but probably out of scope</summary>
I need to check but I think that (with angular) you'd have something like this:

```
interface User {
    username: string
    password: string
}

@Component({
  template: `
    <input type="username" ([ngModel])="user.username">
    <input type="password" ([ngModel])="user.password">
  `
})
class LoginComponent {
  user: User = {username, password}
}
```

I don't see how you'd get `null` in the first place. `undefined` could be a value but I usually remove those before issuing a request. There should be an example with Restangular on the docs. I'm using [rxrest handlers](https://github.com/soyuka/rxrest#handlers) to transform data prior-request.

Also, if `username` should be present, don't you do client-side validation?
</details>

**Comment by vincentchalamon** [comment]
> If username can not be nullable I don't get what response you expect when you send {username: null}, it's invalid. I'd expect 400 status code.

That's also what we expect, but we receive a 500 status code. @daviddlv will paste you the stack trace tomorrow.

> Angular version?

v4.1.2

**Comment by daviddlv** [comment]
Angular version is v4.3.1
I use HttpClientModule and my request is like this :
```
export class User {
  id: string;
  username: string;
  password: string;
  plainPassword: string;
  email: string;
  lastName: string;
  firstName: string;
}

let user: User = new User();
let body = JSON.stringify(this.user);
return this.http.post<User>(`${environment.API_BACKEND_ORIGIN}/users`, body).subscribe(
    data => { this.user = data; }
);
```

The stack trace Symfony is :
```
Uncaught PHP Exception Symfony\Component\Serializer\Exception\UnexpectedValueException: "Expected argument of type "string", "NULL" given" at /app/vendor/symfony/symfony/src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php line 204
Hide trace
{▼
  /app/vendor/symfony/symfony/src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:204: {▶}
  /app/vendor/api-platform/core/src/Serializer/AbstractItemNormalizer.php:115: {▶}
  /app/vendor/api-platform/core/src/Serializer/ItemNormalizer.php:41: {▶}
  /app/vendor/symfony/symfony/src/Symfony/Component/Serializer/Serializer.php:292: {▶}
  /app/vendor/symfony/symfony/src/Symfony/Component/Serializer/Serializer.php:174: {▶}
  /app/vendor/symfony/symfony/src/Symfony/Component/Serializer/Serializer.php:132: {▶}
  /app/vendor/api-platform/core/src/EventListener/DeserializeListener.php:69: {▶}
  /app/var/cache/dev/appDevDebugProjectContainer.php:1592: {▶}
  appDevDebugProjectContainer->{closure}(): {}
  /app/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/WrappedListener.php:112: {▶}
  Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke(): {}
  /app/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:174: {▶}
  /app/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:43: {▶}
  /app/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/TraceableEventDispatcher.php:146: {▶}
  /app/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:129: {▶}
  /app/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:68: {▶}
  /app/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php:171: {▶}
  /app/web/app.php:20: {▶}
}
```

I tried with `@Assert\NotNull` but same error

**Comment by meyerbaptiste** [comment]
IMO, not really a bug but just a bit of logic here.

Indeed, in order to do the validation we need a populated object. However, in order to obtain a populated object, we have to denormalize the received data (i.e., the JSON transformed into a PHP array). The validation is done during the `kernel.view` event and the denormalization during the `kernel.request` event.

If we take your example, your `username` attribute is a strict string type and not nullable (cf. the method signature of your setter and getter, your PHPDoc and your `Column` annotation) and it's these infos that the denormalizer will use to do its job:
```
Symfony\Component\PropertyInfo\Type {
    -builtinType: "string"
    -nullable: false
    -class: null
    -collection: false
    -collectionKeyType: null
    -collectionValueType: null
  }
```

So when you send it `null` and according to the previous infos, it seems logical that the denormalizer has no choice but to tell you `Symfony\Component\Serializer\Exception\UnexpectedValueException`. The denormalizer expects a string but you give it a null value. It's common sense, isn't it?

The HTTP status code 500 may look brutal but you can change it:
```yaml
api_platform:
    exception_to_status:
        Symfony\Component\Serializer\Exception\UnexpectedValueException: !php/const:Symfony\Component\HttpFoundation\Response::HTTP_BAD_REQUEST
```
Another solution is to initialize your model by an empty string in the front.

**Comment by soyuka** [comment]
> Another solution is to initialize your model with an empty string in the front.

Another big :+1:. As I wanted to show in my above angular example, I don't get how you can send `null`, especially if you're working with Typescript, it should enforce you to use either an empty string or `undefined` (`undefined` will not show up in a JSON string sent to the server):

```
> JSON.stringify({foo: undefined, bar: ''})
'{"bar":""}'
```

**Comment by vincentchalamon** [comment]
@meyerbaptiste I agree with you about strict typehint, if we don't allow null value on property and setter, PHP cannot understand it. That's why we also tried allowing null values on class property, getter & setter: we still got the same 500 error. Adding `nullable=true` doctrine option on class property `@ORM\Column` annotation fixed it.

@soyuka In your example, you send `undefined`, but if you send `null` the result is different:
```
> JSON.stringify({foo: null, bar: ''})
'{"foo":null,"bar":""}'
```
I don't know if typescript converts it properly (apparently no cause the API seems to receive `null` instead of `undefined` as property value).

**Comment by soyuka** [comment]
Yes null is stringifyable. But if you've null it's that you really want to send null IMO ^^.

**Comment by vincentchalamon** [comment]
> But if you've null it's that you really want to send null

Like if I want to empty totally a property without sending an empty string? Yes, that should be possible IMO

Seen with @meyerbaptiste, he explained me the DoctrineExtractor has priority to PhpDocExtractor & ReflectionExtractor. I think it should not, specially about the `nullable` property option. PHP class should be decorrelated from Doctrine schema.

**Comment by soyuka** [comment]
If changing decoration priority fixes your issue then :+1:.
